### ISSUE TITLE
One finite field to rule them all

### ISSUE THREAD

**Issue by thofma** [oscar-system/Oscar.jl#1987]
As many of us are painfully reminded from time to time, we have four (actually five) different finite field types. This is not only a burden for developers, but also for users. Due to heroic efforts of the flint maintainers, there is actually one finite field type (used to be known as `fq_default`) that we could use instead. Although it is already hooked up in Nemo, so far we did not adjust the other packages. I hope that we can do the transition as soon as possible (also in view of *the* book). I am opening this issue to track progress and coordinate our efforts:

- [x] Adjust Hecke (WIP, https://github.com/thofma/Hecke.jl/pull/989)
- [x] Tweak the Singular interface (`singular_coeff_ring` produces a generic `n_Field{}` object, which is bad for performance).
- [x] Adjust the GAP interface (I did some small `iso_oscar_gap` tests and it seems to work.)
- [x] #2897
- [x] <s> Get https://github.com/flintlib/flint2/pull/1261 into our flint2 binaries</s> (I worked around it)
- [x] #2192
- [x] Introduce `Native.GF` and `Native.FiniteField` in Nemo and use it in Hecke where necessary.

Once we have this, we can

- [ ] Make the new type the default. This means making `GF` and `FiniteField` always construct the new type.

P.S.: At the moment, to construct the new type, one can use `Nemo._FiniteField` and `Nemo._GF`. 

Feel free to add more things to the list.

**Comment by fingolfin** [comment]
See also issue #867

**Comment by ThomasBreuer** [comment]
@thofma Concerning the GAP interface:
`Nemo._GF(p)` and `Nemo._GF(p, d)` return objects of type `FqField`, and this type is supported by `iso_oscar_gap`.

`iso_gap_oscar` chooses an Oscar field by calling `GF(p)` or `GF(p, d)`.
Thus everything will be fine as soon as `GF` will return the default fields,
but currently a field of the type `Nemo.fpField` or `fqPolyRepField` is used.
Wouldn't it be better to change `iso_gap_oscar` already now to choose the Oscar field by calling `Nemo._GF`?
Then there will be at most a name change in the end, but no change in the behaviour.

**Comment by thofma** [comment]
It would be an interesting experiment to see what breaks when one changes `iso_gap_oscar` as you suggested. Would be great if we could make a draft PR with that change so we can see the CI run.

**Comment by ThomasBreuer** [comment]
@thofma O.k., I will do this.

**Comment by thofma** [comment]
@antonydellavecchia The behavior of the (soon to be default) one finite field, is slightly different. Now one can also construct extensions over arbitrary finite fields. Everything has the same type!

```julia
julia> F, a = Hecke.Nemo._FiniteField(4);

julia> Fx, x = F["x"];

julia> f = x^2 + a*x + 1;

julia> FF, = Hecke.Nemo._FiniteField(f);

julia> defining_polynomial(FF)
x^2 + o*x + 1

julia> coeff(rand(FF), 1)
o
```

Can you check that the serialization can handle this? Happy to talk on slack if there are questions about the internals.

**Comment by antonydellavecchia** [comment]
> @antonydellavecchia The behavior of the (soon to be default) one finite field, is slightly different. Now one can also construct extensions over arbitrary finite fields. Everything has the same type!
> 
> ```julia
> julia> F, a = Hecke.Nemo._FiniteField(4);
> 
> julia> Fx, x = F["x"];
> 
> julia> f = x^2 + a*x + 1;
> 
> julia> FF, = Hecke.Nemo._FiniteField(f);
> 
> julia> defining_polynomial(FF)
> x^2 + o*x + 1
> 
> julia> coeff(rand(FF), 1)
> o
> ```
> 
> Can you check that the serialization can handle this? Happy to talk on slack if there are questions about the internals.

So this will break the current serialization, but I don't think it will be a hard fix

**Comment by thofma** [comment]
Implemented in #3153
